### PR TITLE
feat(bigquery): expose hive partitioning options

### PR DESCRIPTION
**Pull request by shollyman**
This PR resuscitates https://code-review.googlesource.com/c/gocloud/+/43270 
and updates it with additional functionality that was previously not properly exposed
such as RequirePartitionFilter.

This also adds a quick integration test as table creation validates the layout, and issues
an example query that leverages the hive partitioning layout to prune reading.

### CODE REVIEW

File: bigquery/external.go [hunk 1]
```diff
@@ -409,3 +419,66 @@ func bqToBigtableColumn(q *bq.BigtableColumn) (*BigtableColumn, error) {
 	}
 	return b, nil
 }
+
+// HivePartitioningMode is used in conjunction with HivePartitioningOptions.
+type HivePartitioningMode string
+
+var (
```
Comment (hongalex) [inline bigquery/external.go:426]: consider using const here

File: bigquery/external.go [hunk 2]
```diff
@@ -409,3 +419,66 @@ func bqToBigtableColumn(q *bq.BigtableColumn) (*BigtableColumn, error) {
...
+)
+
+// HivePartitioningOptions defines the behavior of Hive partitioning
+// when working with external data.
+type HivePartitioningOptions struct {
+
+	// Mode defines which hive partitioning mode to use when reading data.
+	Mode HivePartitioningMode
+
+	// When hive partition detection is
```
Comment (hongalex) [inline bigquery/external.go:442]: nit: comment formatting looks a bit strange here

**Review by hongalex — APPROVED** [review]

**Comment by hongalex** [comment]
LGTM pending tests green (saw you just pushed a commit fixing the exported var issue )

File: bigquery/external.go (hunk 1 above)
Comment (shollyman) [reply bigquery/external.go]: done

File: bigquery/external.go (hunk 2 above)
Comment (shollyman) [reply bigquery/external.go]: copy/paste from discovery.  updated.

**Pull request merged**